src/org/rascalmpl/library/util/IDEServicesLibrary.java

```diff
@@ -56,8 +58,61 @@ public void registerLanguage(IConstructor language) {
     public void applyDocumentsEdits(IList edits) {
         applyFileSystemEdits(edits);
     }
+
     public void applyFileSystemEdits(IList edits) {
-        services.applyFileSystemEdits(edits);
+        var registry = URIResolverRegistry.getInstance();
+        var vf = IRascalValueFactory.getInstance();
+        edits.stream().map(IConstructor.class::cast).forEach(c -> {
+            try {
+                switch (c.getName()) {
+                    case "removed": {
+                        var file = (ISourceLocation) (c.get("file"));
+                        registry.remove(file.top(), false);
+                        break;
+                    }
+                    case "created": {
+                        var file = (ISourceLocation) (c.get("file"));
+                        if (registry.exists(file)) {
+                            registry.setLastModified(file, System.currentTimeMillis());
+                        } else {
+                            var out = registry.getCharacterWriter(file.top(), "UTF-8", false);
+                            out.write("");
+                            out.close();
+                        }
+                        break;
+                    }
+                    case "renamed": {
+                        var from = (ISourceLocation) (c.get("from"));
+                        var to = (ISourceLocation) (c.get("to"));
+                        registry.rename(from.top(), to.top(), true);
+                        break;
+                    }
+                    case "changed": {
+                        var file = ((ISourceLocation) (c.get("file")));
+                        if (c.has("edits")) {
+                            var textEdits = (IList) (c.get("edits"));
+                            var contents = Prelude.readFile(vf, false, ((ISourceLocation) (c.get("file"))).top(), "UTF-8", false).getValue();
+                            for (var e : textEdits.reverse()) {
+                                var edit = (IConstructor) e;
+                                var range = (ISourceLocation) (edit.get("range"));
+                                var prefix = contents.substring(0, range.getOffset());
+                                var replacement = ((IString) edit.get("replacement")).getValue();
+                                var postfix = contents.substring(range.getOffset() + range.getLength());
```

Comment:
I don't think this )and the other calls on `substring` is correct with surrogate pairs. use `IString::substring` to save ourselves a lot of subtle bugs. (and `IString::concact` etc)